### PR TITLE
Danthb/add cobra cli to sitemap

### PR DESCRIPTION
**Pull request by BarathaAberathne**
### What

I have added cobra cli package to the exist dp-sitemap tool and added generate, update commands using cobra along with the usage messages. 

### How to review

basic generate command example ➜  go run main.go generate --robots-file-path=robot_filesssss.txt --scroll-size=101 --scroll-timeout=2000 --api-url=http://localhostt --sitemap-index=12 --zebedee-url=http://localhost:80821234121334 --sitemap-file-path=sitemap --fake-scroll=true 

basic update command example ➜  go run main.go update --robots-file-path=robot_filesssss.txt --scroll-size=101 --scroll-timeout=2000 --api-url=http://localhostt --sitemap-index=12 --zebedee-url=http://localhost:80821234121334 --sitemap-file-path=sitemap --fake-scroll=true 

### Who can review

Anyone @dionisislolas @lindenmckenzie

### CODE REVIEW

**Review by lindenmckenzie — CHANGES_REQUESTED** [review]
4/5 tests are failing
There's also a lot of commented out code and changes where you're reverting things Dionisi has put in but I can't see the reasons for it, could you both get together to discuss?

**Review by BarathaAberathne — COMMENTED** [review]
Comments are addressed. removed unwanted test file and main file from the sitemapcli folder and moved the function logic into utility folder.

**Comment by lindenmckenzie** [comment]
> Comments are addressed. removed unwanted test file and main file from the sitemapcli folder and moved the function logic into utility folder.

lint and audit are still failing.

File: cmd/generate.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+		cfg, err := config.Get()
+		if err != nil {
+			fmt.Println("Error retrieving config" + err.Error())
+			os.Exit(1)
+		}
+
+		viper.BindPFlags(cmd.Flags()) // Bind Flags with Viper
+
+		// Validate ApiUrl
```
Comment (lindenmckenzie) [inline cmd/generate.go]: Don't think this level of commentary is necessary.

File: cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+var rootCmd = &cobra.Command{
+	Use:   "dp-sitemap",
+	Short: "dp-sitemap is a cli tool to generate sitemaps easily & locally",
+	Long: `dp-sitemap is a comprehensive tool for generating and managing sitemaps of your website. 
+	The tool provides functionality for creating sitemaps from different data sources, updating existing sitemaps, and integrating with external services
+
+Sitemap-cli is a CLI tool for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly.`,
+}
```
Comment (lindenmckenzie) [inline cmd/root.go]: Could indent this better.

File: cmd/root.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	}
+}
+
+func init() {
+	// Here you will define your flags and configuration settings.
+	// Cobra supports persistent flags, which, if defined here,
+	// will be global for your application.
+
+	// rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default s $HOME/.dp-sitemap.yaml)")
+
```
Comment (lindenmckenzie) [inline cmd/root.go]: What's this?

File: cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	err := rootCmd.Execute()
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+func init() {
+	// Here you will define your flags and configuration settings.
+	// Cobra supports persistent flags, which, if defined here,
+	// will be global for your application.
```
Comment (lindenmckenzie) [inline cmd/root.go]: This all looks like commentary copied out of cobra instructions?

File: cmd/update.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			RobotsFilePath:  viper.GetString("robots-file-path"),
+			ApiUrl:          viper.GetString("api-url"),
+			SitemapIndex:    viper.GetString("sitemap-index"),
+			ScrollTimeout:   viper.GetString("scroll-timeout"),
+			ScrollSize:      viper.GetInt("scroll-size"),
+			SitemapPath:     viper.GetString("sitemap-file-path"),
+			ZebedeeUrl:      viper.GetString("zebedee-url"),
+			FakeScroll:      viper.GetBool("fake-scroll"),
+			GenerateSitemap: viper.GetBool("generate-sitemap"),
+			UpdateSitemap:   viper.GetBool("update-sitemap"),
```
Comment (lindenmckenzie) [inline cmd/update.go]: I don't think you need these now you've separated out the commands with cobra?

File: cmd/utilities/utilities.go [hunk 6]
```diff
@@ -172,41 +80,89 @@ func GenerateSitemap(cfg *config.Config, commandline *FlagFields) {
...
 
+func UpdateSitemap(cfg *config.Config, commandLine *FlagFields) {
+
+	cfg, err := config.Get()
+	if err != nil {
+		fmt.Println("Error retrieving config" + err.Error())
+		os.Exit(1)
+	}
+
+	if commandLine.UpdateSitemap {
```
Comment (lindenmckenzie) [inline cmd/utilities/utilities.go]: I don't think this is necessary now? Specifically the boolean `commandLine.UpdateSitemap`

File: cmd/utilities/utilities.go [hunk 7]
```diff
@@ -172,41 +80,89 @@ func GenerateSitemap(cfg *config.Config, commandline *FlagFields) {
...
+	saveErr := store.SaveFile(commandline.RobotsFilePath, strings.NewReader(body))
 	if saveErr != nil {
 		fmt.Println("failed to save file")
 		return
 	}
 }
 
+func UpdateSitemap(cfg *config.Config, commandLine *FlagFields) {
+
+	cfg, err := config.Get()
```
Comment (lindenmckenzie) [inline cmd/utilities/utilities.go]: You're passing in the config but then getting it again?

File: go.mod [hunk 8]
```diff
@@ -2,9 +2,6 @@ module github.com/ONSdigital/dp-sitemap
 
 go 1.20
 
-// to fix: [CVE-2023-32731] CWE-Other
-replace google.golang.org/grpc => google.golang.org/grpc v1.55.0
-
```
Comment (lindenmckenzie) [inline go.mod]: This is probably what's causing the audit failure.

File: main.go [hunk 9]
```diff
@@ -25,13 +26,16 @@ var (
 )
 
 func main() {
+	cmd.Execute()
+
```
Comment (lindenmckenzie) [inline main.go]: This looks like you're forcing the tool to be command line?

**Review by lindenmckenzie — CHANGES_REQUESTED** [review]
Some further comments on the code itself.

File: cmd/root.go (hunk 4 above)
Comment (BarathaAberathne) [reply cmd/root.go]: these lines are auto generate with the template. will be removed.

File: cmd/root.go (hunk 3 above)
Comment (BarathaAberathne) [reply cmd/root.go]: Unwanted lines. will be removed.

File: go.mod (hunk 8 above)
Comment (dionisislolas) [reply go.mod]: replace github.com/hashicorp/consul/api => github.com/hashicorp/consul/api v1.11.5

I tried an older version of this dependency and it is passing the audit. It is mentioned in the description of the vulnerability that it is fixed in 1.9.17, 1.10.10, and 1.11.5.

File: cmd/update.go (hunk 5 above)
Comment (BarathaAberathne) [reply cmd/update.go]: This is the standard way to use cobra with viper. I have use viper to do a validation first and then assigned the validated flags into our FlagField structure.

File: main.go (hunk 9 above)
Comment (BarathaAberathne) [reply main.go]: This was in the documentation, we have to add this cmd.execute() into our main method in order to make the commands work.

**Review by BarathaAberathne — COMMENTED** [review]
All the comments addressed. Some changes remains same as they are the standard ways i found from documentations.

**Review by BarathaAberathne — COMMENTED** [review]
Dependency updated

**Review by BarathaAberathne — COMMENTED** [review]
All comments addressed.

File: cmd/commands.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	cmd.Flags().String("robots-file-path", "test_robots.txt", "path to robots file")
+	cmd.Flags().String("sitemap-file-path", "test_sitemap", "path to sitemap file")
+	cmd.Flags().String("api-url", "http://localhost", "elastic search api url")
+	cmd.Flags().String("zebedee-url", "http://localhost:8082", "zebedee url")
+	cmd.Flags().String("sitemap-index", "1", "OPENSEARCH_SITEMAP_INDEX")
+	cmd.Flags().String("scroll-timeout", "2000", "OPENSEARCH_SCROLL_TIMEOUT")
+	cmd.Flags().Int("scroll-size", 10, "OPENSEARCH_SCROLL_SIZE")
+	cmd.Flags().Bool("fake-scroll", true, "enable fake scroll")
```
Comment (lindenmckenzie) [inline cmd/commands.go]: You could set these as root command flags rather than duplicating them.

File: cmd/utilities/utilities.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+	)
+
+	// Generating sitemap
+	genErr := generator.MakeFullSitemap(context.Background())
+	if genErr != nil {
+		fmt.Println("Error writing sitemap file", genErr.Error())
+		return
+	}
+
+	GenerateRobotFile(cfg, commandline)
```
Comment (lindenmckenzie) [inline cmd/utilities/utilities.go]: Would be sensible to separate this out and not have the sitemap creation call the robots creation I think?

File: cmd/utilities/utilities.go [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func getContent() (*event.ContentPublished, error) {
+	content := &event.ContentPublished{}
+	fmt.Print("Please enter URI: ")
+	text, err := getText()
+	if err != nil {
+		return nil, err
+	}
+	content.URI = *text
```
Comment (lindenmckenzie) [inline cmd/utilities/utilities.go:153]: I think this pattern seems messy - mostly the re-use of `text`

File: cmd/utilities/utilities.go [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+func getText() (*string, error) {
+	reader := bufio.NewReader(os.Stdin)
+	text, err := reader.ReadString('\n')
+	if err != nil {
+		return nil, err
+	}
+	// we convert CRLF to LF
+	text = strings.Replace(text, "\n", "", -1)
+	return &text, nil
```
Comment (lindenmckenzie) [inline cmd/utilities/utilities.go:195]: could use bufio.Scanner here instead maybe?

**Review by lindenmckenzie — DISMISSED** [review]
Made a few comments - happy to let it merge as is or if you want to make changes please do.

**Review by lindenmckenzie — APPROVED** [review]

**Pull request merged**